spring-kafka/src/test/java/org/springframework/kafka/support/resolver/ResolvableTypeResolverTests.java

```diff
@@ -0,0 +1,165 @@
+
+/**
+ * @author Scruel Tao
+ */
+public class ResolvableTypeResolverTests {
+	private static final String TEXT = "kafka";
+
+	@Test
+	void jsonFoo() throws Exception {
+		InvocableHandlerMethod method = getFooInvocableHandlerMethod();
+		String s = new ObjectMapper().writeValueAsString(new Foo().setText(TEXT));
+		method.invoke(new GenericMessage<>(s));
+	}
+
+	@Test
+	void bytesToFoo() throws Exception {
+		InvocableHandlerMethod method = getFooInvocableHandlerMethod();
+		String s = new ObjectMapper().writeValueAsString(new Foo().setText(TEXT));
+		method.invoke(new GenericMessage<>(s.getBytes(StandardCharsets.UTF_8)));
+	}
+
+	public void onMessageWithFoo(@ResolvableType Foo payload) {
+		assertThat(payload.getText()).isEqualTo(TEXT);
+	}
+
+	@Test
+	void stringToFooForceJson() throws Exception {
+		final InvocableHandlerMethod method = getFooInvocableHandlerMethod();
+		final String s = new ObjectMapper().writeValueAsString(new Foo().setText(TEXT));
+		final Map<String, Object> contentType = Map.of(MessageHeaders.CONTENT_TYPE, "application/xml");
+		// No converters can process xml media type, so will cause MessageConversionException.
+		assertThatThrownBy(() ->
+				method.invoke(new GenericMessage<>(s, contentType)))
+				.isExactlyInstanceOf(MessageConversionException.class);
+		// Ignore the header, and force using own converter.
+		InvocableHandlerMethod forceMethod = getForceJsonFooInvocableHandlerMethod();
+		forceMethod.invoke(new GenericMessage<>(s, contentType));
+	}
+
+	public void onForceJsonMessageWithFoo(@ResolvableType(force = true) Foo payload) {
+		assertThat(payload.getText()).isEqualTo(TEXT);
+	}
+
+	@NotNull
+	private InvocableHandlerMethod getFooInvocableHandlerMethod() throws NoSuchMethodException {
+		MessageHandlerMethodFactory factory = getMessageHandlerMethodFactory();
+		return factory.createInvocableHandlerMethod(this, getClass().getDeclaredMethod(
+				"onMessageWithFoo", Foo.class));
+	}
+
+	@NotNull
+	private InvocableHandlerMethod getForceJsonFooInvocableHandlerMethod() throws NoSuchMethodException {
+		MessageHandlerMethodFactory factory = getMessageHandlerMethodFactory();
+		return factory.createInvocableHandlerMethod(this, getClass().getDeclaredMethod(
+				"onForceJsonMessageWithFoo", Foo.class));
+	}
+
+	@SuppressWarnings("rawtypes")
+	private MessageHandlerMethodFactory getMessageHandlerMethodFactory() {
+		KafkaListenerAnnotationBeanPostProcessor bpp = new KafkaListenerAnnotationBeanPostProcessor<>();
+		beanPostProcessorConfig(bpp);
+		return bpp.getMessageHandlerMethodFactory();
+	}
+
+	@SuppressWarnings("rawtypes")
+	private void beanPostProcessorConfig(KafkaListenerAnnotationBeanPostProcessor processor) {
+		processor.setBeanFactory(new KafkaBeanFactory());
+		try {
+			processor.afterSingletonsInstantiated();
+		} catch (NoSuchBeanDefinitionException ignore) {
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	static class KafkaBeanFactory extends DefaultListableBeanFactory {
```

Comment:
It used to invoke `setCustomMethodArgumentResolvers`, because all relevant methods are private, so I choose this way to achieve the goal.

Reply:
OK. I see your point. It looks like you try mock some infrastructure for `spring-kafka`.
Let it be for now!
But I think it can be simplified anyway. See `HeaderMethodArgumentResolverTests` for example.

Reply:
Ok, thanks, I am a little busy right now, so will check this tomorrow.

Reply:
@artembilan Checked, and the code already based on `HeaderMethodArgumentResolverTests`, which is not contained the logic I need.